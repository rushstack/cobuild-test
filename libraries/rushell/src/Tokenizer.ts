// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TextRange } from './TextRange';
import { ParseError } from './ParseError';

export enum TokenKind {
  // One or more spaces/tabs
  Spaces,

  // A single newline sequence such as CRLF or LF
  NewLine,

  // A general character without any special meaning
  OtherCharacter,

  // A sequence of characters that doesn't contain any symbols with special meaning
  // Characters can be escaped, in which case the Token.text may differ from the
  // Token.range.toString()
  Text,

  // The "&&" operator, which executes the following command only if the preceding command
  // succeeded (i.e. returned a zero exit code).
  AndIf,

  // A double-quoted string which can do variable expansions
  DoubleQuotedText,

  // A dollar sign followed by an environment variable name
  DollarVariable,

  // The end of the input string
  EndOfInput
}

export class Token {
  public readonly kind: TokenKind;
  public readonly range: TextRange;

  /**
   * The extracted content, which depends on the type:
   *
   * Text: The unescaped content
   * DoubleQuotedText: The unescaped contents inside the quotes.
   * DollarVariable: The variable name without the "$"
   */
  public readonly text: string;

  public constructor(kind: TokenKind, range: TextRange, text?: string) {
    this.kind = kind;
    this.range = range;
    this.text = text === undefined ? this.range.toString() : text;
  }

  public toString(): string {
    return this.text;
  }
}

const textCharacterRegExp: RegExp = /[a-z0-9_\\]/i;
const startVariableCharacterRegExp: RegExp = /[a-z_]/i;
const variableCharacterRegExp: RegExp = /[a-z0-9_]/i;

export class Tokenizer {
  public readonly input: TextRange;
  private _currentIndex: number;

  public constructor(input: TextRange | string) {
    if (typeof input === 'string') {
      this.input = TextRange.fromString(input);
    } else {
      this.input = input;
    }
    this._currentIndex = this.input.pos;
  }

  private static _isSpace(c: string | undefined): boolean {
    // You can empirically test whether shell treats a given character as whitespace like this:
    // echo $(echo -e a '\u0009' b)
    // If you get "a b" it means the tab character (Unicode 0009) is being collapsed away.
    // If you get "a   b" then the invisible character is being padded like a normal letter.
    return c === ' ' || c === '\t';
  }

  public get currentIndex(): number {
    return this._currentIndex;
  }

  public readToken(): Token {
    const input: TextRange = this.input;

    const startIndex: number = this._currentIndex;
    const firstChar: string | undefined = this._peekCharacter();

    // Reached end of input yet?
    if (firstChar === undefined) {
      return new Token(TokenKind.EndOfInput, TextRange.empty);
    }

    // Is it a sequence of whitespace?
    if (Tokenizer._isSpace(firstChar)) {
      this._readCharacter();

      while (Tokenizer._isSpace(this._peekCharacter())) {
        this._readCharacter();
      }

      return new Token(TokenKind.Spaces, input.getNewRange(startIndex, this._currentIndex));
    }

    // Is it a newline?
    if (firstChar === '\r') {
      this._readCharacter();
      if (this._peekCharacter() === '\n') {
        this._readCharacter();
      }
      return new Token(TokenKind.NewLine, input.getNewRange(startIndex, this._currentIndex));
    } else if (firstChar === '\n') {
      this._readCharacter();
      return new Token(TokenKind.NewLine, input.getNewRange(startIndex, this._currentIndex));
    }

    // Is it a double-quoted string?
    if (firstChar === '"') {
      this._readCharacter(); // consume the opening quote

      let text: string = '';
      let c: string | undefined = this._peekCharacter();
      while (c !== '"') {
        if (c === undefined) {
          throw new ParseError(
            'The double-quoted string is missing the ending quote',
            input.getNewRange(startIndex, this._currentIndex)
          );
        }
        if (c === '\r' || c === '\n') {
          throw new ParseError(
            'Newlines are not supported inside strings',
            input.getNewRange(this._currentIndex, this._currentIndex + 1)
          );
        }

        // NOTE: POSIX says that backslash acts as an escape character inside a double-quoted string
        // ONLY if followed by certain other characters.  For example, yes for "a\$" but no for "a\t".
        // Whereas Dash says yes for "a\t" but no for "a\q".  And then Bash says yes for "a\t".
        // This goes against Rushell's goal of being intuitive:  Nobody should have to memorize a list
        // of alphabet letters that cannot be escaped.  So we just say that backslash is *always* an
        // escape character inside a double-quoted string.
        //
        // NOTE: Dash interprets "\t" as a tab character, but Bash does not.
        if (c === '\\') {
          this._readCharacter(); // discard the backslash
          if (this._peekCharacter() === undefined) {
            throw new ParseError(
              'A backslash must be followed by another character',
              input.getNewRange(this._currentIndex, this._currentIndex + 1)
            );
          }
          // Add the escaped character
          text += this._readCharacter();
        } else {
          text += this._readCharacter();
        }

        c = this._peekCharacter();
      }
      this._readCharacter(); // consume the closing quote

      return new Token(TokenKind.DoubleQuotedText, input.getNewRange(startIndex, this._currentIndex), text);
    }

    // Is it a text token?
    if (textCharacterRegExp.test(firstChar)) {
      let text: string = '';
      let c: string | undefined = firstChar;
      do {
        if (c === '\\') {
          this._readCharacter(); // discard the backslash
          if (this._peekCharacter() === undefined) {
            throw new ParseError(
              'A backslash must be followed by another character',
              input.getNewRange(this._currentIndex, this._currentIndex + 1)
            );
          }
          // Add the escaped character
          text += this._readCharacter();
        } else {
          text += this._readCharacter();
        }

        c = this._peekCharacter();
      } while (c && textCharacterRegExp.test(c));

      return new Token(TokenKind.Text, input.getNewRange(startIndex, this._currentIndex), text);
    }

    // Is it a dollar variable?  The valid environment variable names are [A-Z_][A-Z0-9_]*
    if (firstChar === '$') {
      this._readCharacter();

      let name: string = this._readCharacter() || '';
      if (!startVariableCharacterRegExp.test(name)) {
        throw new ParseError(
          'The "$" symbol must be followed by a letter or underscore',
          input.getNewRange(startIndex, this._currentIndex)
        );
      }

      let c: string | undefined = this._peekCharacter();
      while (c && variableCharacterRegExp.test(c)) {
        name += this._readCharacter();
        c = this._peekCharacter();
      }
      return new Token(TokenKind.DollarVariable, input.getNewRange(startIndex, this._currentIndex), name);
    }

    // Is it the "&&" token?
    if (firstChar === '&') {
      if (this._peekCharacterAfter() === '&') {
        this._readCharacter();
        this._readCharacter();
        return new Token(TokenKind.AndIf, input.getNewRange(startIndex, this._currentIndex));
      }
    }

    // Otherwise treat it as an "other" character
    this._readCharacter();
    return new Token(TokenKind.OtherCharacter, input.getNewRange(startIndex, this._currentIndex));
  }

  public readTokens(): Token[] {
    const tokens: Token[] = [];
    let token: Token = this.readToken();
    while (token.kind !== TokenKind.EndOfInput) {
      tokens.push(token);
      token = this.readToken();
    }
    return tokens;
  }

  /**
   * Retrieve the next character in the input stream.
   * @returns a string of length 1, or undefined if the end of input is reached
   */
  private _readCharacter(): string | undefined {
    if (this._currentIndex >= this.input.end) {
      return undefined;
    }
    return this.input.buffer[this._currentIndex++];
  }

  /**
   * Return the next character in the input stream, but don't advance the stream pointer.
   * @returns a string of length 1, or undefined if the end of input is reached
   */
  private _peekCharacter(): string | undefined {
    if (this._currentIndex >= this.input.end) {
      return undefined;
    }
    return this.input.buffer[this._currentIndex];
  }

  /**
   * Return the character after the next character in the input stream, but don't advance the stream pointer.
   * @returns a string of length 1, or undefined if the end of input is reached
   */
  private _peekCharacterAfter(): string | undefined {
    if (this._currentIndex + 1 >= this.input.end) {
      return undefined;
    }
    return this.input.buffer[this._currentIndex + 1];
  }
}
