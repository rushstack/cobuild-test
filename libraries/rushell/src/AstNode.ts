// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Token } from './Tokenizer';
import { TextRange } from './TextRange';

export const enum AstKind {
  None = 'None',
  Script = 'Script',
  AndIf = 'AndIf',
  Command = 'Command',
  CompoundWord = 'CompoundWord',
  VariableExpansion = 'VariableExpansion',
  Text = 'Text'
}

/**
 * Base class for all AST nodes.
 */
export abstract class AstBaseNode {
  public readonly kind: AstKind = AstKind.None;
  public range: TextRange | undefined;

  /**
   * Returns a diagnostic dump of the tree, showing the prefix/suffix/separator for
   * each node.
   */
  public getDump(indent: string = ''): string {
    const nestedIndent: string = indent + '  ';
    let result: string = indent + `- ${this.kind}:\n`;

    const dumpText: string | undefined = this.getDumpText();
    if (dumpText) {
      result += nestedIndent + 'Value=' + JSON.stringify(dumpText) + '\n';
    }

    const fullRange: TextRange = this.getFullRange();
    if (!fullRange.isEmpty()) {
      result += nestedIndent + 'Range=' + JSON.stringify(fullRange.toString()) + '\n';
    }

    const childNodes: AstNode[] = this.getChildNodes();
    for (const child of childNodes) {
      result += child.getDump(nestedIndent);
    }

    return result;
  }

  public getChildNodes(): AstNode[] {
    const nodes: AstNode[] = [];
    this.collectChildNodesInto(nodes);
    return nodes;
  }

  public getFullRange(): TextRange {
    if (this.range) {
      return this.range;
    }

    let encompassingRange: TextRange = TextRange.empty;

    for (const child of this.getChildNodes()) {
      encompassingRange = encompassingRange.getEncompassingRange(child.getFullRange());
    }

    return encompassingRange;
  }

  protected abstract collectChildNodesInto(nodes: AstNode[]): void;

  protected getDumpText(): string | undefined {
    return undefined;
  }
}

/**
 * Represents a complete script that can be executed.
 */
export class AstScript extends AstBaseNode {
  public readonly kind: AstKind.Script = AstKind.Script;

  public body: AstNode | undefined;

  /** @override */
  protected collectChildNodesInto(nodes: AstNode[]): void {
    if (this.body) {
      nodes.push(this.body);
    }
  }
}

/**
 * Represents the "&&" operator, which is used to join two individual commands.
 */
export class AstAndIf extends AstBaseNode {
  public readonly kind: AstKind.AndIf = AstKind.AndIf;

  /**
   * The command that executes first, and always.
   */
  public firstCommand: AstCommand | undefined;

  /**
   * The command that executes second, and only if the first one succeeds.
   */
  public secondCommand: AstCommand | undefined;

  /** @override */
  protected collectChildNodesInto(nodes: AstNode[]): void {
    if (this.firstCommand) {
      nodes.push(this.firstCommand);
    }
    if (this.secondCommand) {
      nodes.push(this.secondCommand);
    }
  }
}

/**
 * Represents a command.  For example, the name of an executable to be started.
 */
export class AstCommand extends AstBaseNode {
  public readonly kind: AstKind.Command = AstKind.Command;

  public commandPath: AstCompoundWord | undefined;
  public arguments: AstCompoundWord[] = [];

  /** @override */
  protected collectChildNodesInto(nodes: AstNode[]): void {
    if (this.commandPath) {
      nodes.push(this.commandPath);
    }
    nodes.push(...this.arguments);
  }
}

/**
 * Represents a compound word, e.g. "--the-thing" or "./the/thing".
 */
export class AstCompoundWord extends AstBaseNode {
  public readonly kind: AstKind.CompoundWord = AstKind.CompoundWord;

  public readonly parts: AstNode[] = [];

  /** @override */
  protected collectChildNodesInto(nodes: AstNode[]): void {
    nodes.push(...this.parts);
  }
}

/**
 * Represents an environment variable expansion expression, e.g. "${VARIABLE}"
 */
export class AstVariableExpansion extends AstBaseNode {
  public readonly kind: AstKind.VariableExpansion = AstKind.VariableExpansion;

  /** @override */
  protected collectChildNodesInto(nodes: AstNode[]): void {
    // no children
  }
}

/**
 * Represents some plain text.
 */
export class AstText extends AstBaseNode {
  public readonly kind: AstKind.Text = AstKind.Text;

  public token: Token | undefined;

  /** @override */
  protected collectChildNodesInto(nodes: AstNode[]): void {
    // no children
  }

  /** @override */
  protected getDumpText(): string | undefined {
    if (this.token) {
      return this.token.text;
    }
    return undefined;
  }
}

export type AstNode = AstScript | AstAndIf | AstCommand | AstCompoundWord | AstVariableExpansion | AstText;
