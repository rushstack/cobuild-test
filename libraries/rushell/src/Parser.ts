// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ParseError } from './ParseError';
import { Tokenizer, Token, TokenKind } from './Tokenizer';
import { AstNode, AstScript, AstCommand, AstCompoundWord, AstText } from './AstNode';

export class Parser {
  private readonly _tokenizer: Tokenizer;
  private _peekedToken: Token | undefined;

  public constructor(tokenizer: Tokenizer) {
    this._tokenizer = tokenizer;
    this._peekedToken = undefined;
  }

  public parse(): AstScript {
    const script: AstScript = new AstScript();

    const startingToken: Token = this._peekToken();

    const astCommand: AstCommand | undefined = this._parseCommand();

    if (!astCommand) {
      throw new ParseError('Expecting a command', startingToken.range);
    }

    const nextToken: Token = this._peekToken();

    if (nextToken.kind !== TokenKind.EndOfInput) {
      throw new ParseError(`Unexpected token: ${TokenKind[nextToken.kind]}`, nextToken.range);
    }

    script.body = astCommand;

    return script;
  }

  private _parseCommand(): AstCommand | undefined {
    this._skipWhitespace();

    const startingToken: Token = this._peekToken();

    const command: AstCommand = new AstCommand();
    command.commandPath = this._parseCompoundWord();
    if (!command.commandPath) {
      throw new ParseError('Expecting a command path', startingToken.range);
    }

    while (this._skipWhitespace()) {
      const compoundWord: AstCompoundWord | undefined = this._parseCompoundWord();
      if (!compoundWord) {
        break;
      }
      command.arguments.push(compoundWord);
    }

    return command;
  }

  private _parseCompoundWord(): AstCompoundWord | undefined {
    const compoundWord: AstCompoundWord = new AstCompoundWord();

    for (;;) {
      const node: AstNode | undefined = this._parseText();
      if (!node) {
        break;
      }
      compoundWord.parts.push(node);
    }

    if (compoundWord.parts.length === 0) {
      // We didn't parse a word
      return undefined;
    }

    return compoundWord;
  }

  private _parseText(): AstText | undefined {
    const token: Token = this._peekToken();

    if (token.kind === TokenKind.Text) {
      this._readToken();

      const astText: AstText = new AstText();
      astText.token = token;
      astText.range = token.range;
      return astText;
    }

    return undefined;
  }

  /**
   * Skips any whitespace tokens.  Returns true if any whitespace was actually encountered.
   */
  private _skipWhitespace(): boolean {
    let sawWhitespace: boolean = false;
    while (this._peekToken().kind === TokenKind.Spaces) {
      this._readToken();
      sawWhitespace = true;
    }
    if (this._peekToken().kind === TokenKind.EndOfInput) {
      sawWhitespace = true;
    }
    return sawWhitespace;
  }

  private _readToken(): Token {
    if (this._peekedToken) {
      const token: Token = this._peekedToken;
      this._peekedToken = undefined;
      return token;
    } else {
      return this._tokenizer.readToken();
    }
  }

  private _peekToken(): Token {
    if (!this._peekedToken) {
      this._peekedToken = this._tokenizer.readToken();
    }
    return this._peekedToken;
  }
}
