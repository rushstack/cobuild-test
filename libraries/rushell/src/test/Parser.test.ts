// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Tokenizer } from '../Tokenizer';
import { Parser } from '../Parser';
import { AstScript } from '../AstNode';

function escape(s: string): string {
  return s.replace(/\n/g, '[n]').replace(/\r/g, '[r]').replace(/\t/g, '[t]').replace(/\\/g, '[b]');
}

function matchSnapshot(input: string): void {
  const tokenizer: Tokenizer = new Tokenizer(input);
  const parser: Parser = new Parser(tokenizer);
  const result: AstScript = parser.parse();
  expect({
    input: escape(tokenizer.input.toString()),
    tree: '\n' + result.getDump()
  }).toMatchSnapshot();
}

function matchErrorSnapshot(input: string): void {
  const tokenizer: Tokenizer = new Tokenizer(input);
  const parser: Parser = new Parser(tokenizer);
  let error: Error | undefined = undefined;
  try {
    parser.parse();
  } catch (e) {
    error = e as Error;
  }
  expect({
    input: escape(tokenizer.input.toString()),
    reportedError: error
  }).toMatchSnapshot();
}

test('00: basic inputs', () => {
  matchSnapshot('command arg1 arg2');
});

test('01: basic errors', () => {
  matchErrorSnapshot('@bad');
  matchErrorSnapshot('command @bad');
});
