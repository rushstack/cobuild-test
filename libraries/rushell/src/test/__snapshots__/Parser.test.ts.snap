// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`00: basic inputs 1`] = `
Object {
  "input": "command arg1 arg2",
  "tree": "
- Script:
  Range=\\"command arg1 arg2\\"
  - Command:
    Range=\\"command arg1 arg2\\"
    - CompoundWord:
      Range=\\"command\\"
      - Text:
        Value=\\"command\\"
        Range=\\"command\\"
    - CompoundWord:
      Range=\\"arg1\\"
      - Text:
        Value=\\"arg1\\"
        Range=\\"arg1\\"
    - CompoundWord:
      Range=\\"arg2\\"
      - Text:
        Value=\\"arg2\\"
        Range=\\"arg2\\"
",
}
`;

exports[`01: basic errors 1`] = `
Object {
  "input": "@bad",
  "reportedError": [Error: (1,1): Expecting a command path],
}
`;

exports[`01: basic errors 2`] = `
Object {
  "input": "command @bad",
  "reportedError": [Error: (1,9): Unexpected token: OtherCharacter],
}
`;
